LDAR_Sim/model_code/input_mapper_v1.py

```diff
@@ -0,0 +1,100 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# MIT License for more details.
+
+# You should have received a copy of the MIT License
+# along with this program.  If not, see <https://opensource.org/licenses/MIT>.
+#
+# ------------------------------------------------------------------------------
+
+def input_mapper_v1(parameters):
+    """Function to perform all input mapping from version 1.0 parameter files to presently compliant parameter files. This is
+    necessary to ensure reverse compatibility - to allow older parameter files to run properly.
+
+    This function is a series of hooks that fall into several categories:
+    1) Construction hooks: these hooks construct newer parameters from older parameters. This is necessary if parameters
+    were depreciated in the newest version of the model, but relied upon in older versions. For example, newer versions
+    of the program may simply allow specification of a start and end date, but not include specification of the
+    number of timesteps. To allow this, a construction hook could create start and end date from the number of
+    timesteps using any necessary rules to maintain compatibility.
+
+    2) Destruction hooks: these hooks destruct parameters that are no longer in the default set of parameters in the
+    model. To continue the above example, if the present version of the model does not use number of timesteps, and
+    the number of timesteps is no longer specified in the default parameter file - that key should be removed. The issue
+    here is without destruction of depreciated parameters - this parameter file will fail validation. Of course,
+    destruction must be called AFTER construction so the information in the depreciated parameters is used to map
+    to the new parameters before being deleted!
+
+    3) Recalculation hooks: these hooks recalculate variables or map the values. For example, if the units have
+    changed, but the key has not - these hooks are necessary to recalculate the new units to be compliant with the
+    present version of the model. If the older version of the model had a spelling mistake in a string specification,
+    this type of hook would map the parameters. Several examples:
+
+    # map a old spelling mistake properly - the present version of LDAR Sim has corrected the spelling mistake but
+    # old parameter files continue to use the misspelled value
+    if parameters['calculation_method'] == 'categoricle':
+        parameters['calculation_method'] = 'categorical'
+
+    # map the leak rate parameter from one unit to another, the newest version uses a different unit, where 0.003332 is
+    # the conversion factor
+    parameters['leak_rate] = parameters['leak_rate'] * 0.003332
+
+    4) Key-change hooks: these hooks change the key name to a new key name to map the old name to the new name.
+
```

Comment:
@tarcadius @soroushojagh - here is a synopsis of the types of changes that must be specifically dealt with to map parameters from some older version to the version 2.0 specification. 

Many new parameter changes will require an entry here to allow @soroushojagh 's parameter database to map properly. If we choose to support @soroushojagh 's parameter database, which I think we should, we need to do some work here once we figure out the specification @soroushojagh  has used. This mapper will need to be maintained from now until we collectively choose to not support the older specification.

Reply:
Thanks @tbarchyn, 

For version 2.0, if changes are about values of different parameters, it can be easily handled for the Web API DB. 

Example: 

Changes form `'LPR': 0.0065 `  to `'LPR': 0.0085 `

In another scenario, if some new parameters add to the old version of input parameters, again they can be easily added to the DB. 

**My concern:** some structural changes happen in the input parameters. For example, 

`'MDL': [0.01275, 2.78e-6],` 

Changes to: 

`'Parameters1': 0.01275`

`'Parameter2': , 2.78e-6`

Then, it should be time-consuming for me on the Web API side. 

In conclusion, I don't think I need a mapper for changing values or adding new parameters from V1 to V2. But, for structural changes, I definitely need a mapper or re-design of the DB.

Reply:
There have been structural changes to the parameters and will continue to be new changes since when you built the input parameter database.

What branch of LDAR-Sim are you using for the web app right now??

Yes, an input mapper is required. It is already required for v1 files (refer to this file we are commenting on),

What I'm hoping you can provide is a branch that defines the structure of the parameters you have built the database with so I can build at least the start of the mapper. 

If possible, could you provide a branch that defines the version of LDAR sim that works with the web app.

Thanks.

Reply:
I designed the DB based on input parameters in IMTS_Private: `/master/LDAR_Sim/inputs_template`.

Reply:
OK, your branch is:

https://github.com/IM3S/LDAR_Sim/tree/Soroush_input_database

All changes to input parameters from that point forward are going to need to map from that branch to whatever the latest version is. Thanks. 

This is a priority as there are PRs that are going to change the parameters and structure of those parameters and we need to capture the structure and defaults you are using BEFORE those changes are merged - otherwise you will have a lot of unnecessary work to do!